app/assets/javascripts/discourse/models/topic-tracking-state.js

```diff
@@ -132,6 +141,28 @@ const TopicTrackingState = EmberObject.extend({
     });
   },
 
+  trackMutedTopics(data) {
+    if (!this.currentUser) {
+      return [];
+    }
+    let mutedTopics = this.currentUser.muted_topics || [];
+    let now = Date.now();
+
+    if (data.message_type === "muted") {
+      mutedTopics = mutedTopics.concat({
+        topicId: data.topic_id,
+        createdAt: now
+      });
+    } else {
+      mutedTopics = mutedTopics.filter(
```

Comment:
shouldn't it run this unconditionally? always do a cleanup run when you track a new muted topic?

Reply:
I don't think it is necessary. When you mute topic, you will receive 2 messages. First "muted-topic" will just add that topic to an array of muted topics. Second message "latest", will filter muted topics and decide if the message should be processed or not. 

If we filter on "muted" message as well we will do it twice in a very short time and "latest" is a more important message because if you unmute topic you will not receive "muted" message anymore.

Reply:
aha... ok lets refactor a bit, I feel this method is doing a bit too much work...

```
if (data.message_type == "muted") {
   tracker.trackMutedTopic(data.topic_id)
   return;
}

pruneOldMutedTopics();

if (isMutedTopic(data.topic_id)) {
    return;
}


```

a lot more verbose, but significantly less full of side effects so it is far simpler to reason about